### PR TITLE
Enable 2d Segmentation

### PR DESCRIPTION
**Pull request by kbressem**
# Summary
This PR adds support for 2d segmentation. 

## Enhanced viewing
Enhances the interactive viewer to support 2D and 3D data. The viewer can display grayscale images (standard cmap is "bone"), 2D RGB images (`mode = "RGB"` when viewer is initialized), and greyscale 3D images. As labels (`y`) an array (`torch.Tensor`, `np.ndarray`, a string or a number is supported). Additionally, a prediction can be provided alongside the label to visualize correct/wrong examples. An optional description (string) can also be added to each plot. 

All three viewers (`BasicViewer`, `DicomExplorer` and `ListViewer`) support the new functionalities. 

image

In the above image "A Chest X-Ray", "Another X-Ray", "A CT", "Another CT" are file descriptions passed manually. Useful for sharing code (or adding the filename). 
"There is Pneumonia" and "There is Covid" are labels passed as `y` to the viewer. 

Multiple tests for the viewers have been added as well. 

## 2d support for `show_batch`
Show batch can now display 2d, 2d RGB and 3d data. Multichannel data (2d or 3d) is split into single instances and the label is duplicated across all instances

## Increase speed of `show_batch`
`show_batch` is now more than 10 times faster. In the previous implementation, each time a batch was drawn, the data loader first spawned multiple subprocesses, which took considerable time. Now only one data item is drawn, making the function much faster. The larger the transforms pipeline, the more noticeable this should be. 

## Add new test data
Three new 2d images and corresponding labels were added. On purpose, I have made them in different formats (`nrrd`, `nii.gz`, `png`).

## 2d segmentation
SegmentationTrainer now accepts 2d data as input. Config can remain largely unchanged for this. 

## Preparations for classification functionalities
I already started preparations for #41: 
- The config has been renamed and a new classification config was added.
- New data CSV's with classification labels were added. 
- Viewers support labels and masks
- model.py accepts classification networks and provides a workaround for inconsistent argument names in `ResNet`

### CODE REVIEW

File: tests/test_viewer.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        # pyplot windows will open and close during this test in fast succession.
+        # This might lead to the windows blocking each other and the test to fail.
+        # With plt.close("all") and plt.pause(0.001) as well as plt.ion() at the beginning of the file
+        # it seems to work. If tests should fail on another machine, plt.show(block=False) might need to be
+        # added to the viewer classes.
+
+        basic_viewer = BasicViewer(x=image, y=label, prediction=prediction, description=description, mode=mode)
+        basic_viewer.show()
+        plt.close("all")
+        plt.pause(0.001)
```
Comment (markus-hinsche) [inline tests/test_viewer.py:62]: When running the tests, they work, but this is quite irritating. 

I wonder what we want to unit-test here anyway:
* Currently this is a smoke test
* Maybe `basic_viewer.show()` is enough without `plt.show()`: This way we would smoke test the code but not test matplotlib displaying

File: tests/test_viewer.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+        if label is not None:
+            label = [label, label]
+        if prediction is not None:
+            prediction = [prediction, prediction]
+        if description is not None:
+            description = [description, description]
+
+        list_viewer = ListViewer(x=[image, image], y=label, prediction=prediction, description=description, mode=mode)
+        list_viewer.show()
```
Comment (markus-hinsche) [inline tests/test_viewer.py:77]: Is there a sample tutorial script to run when I want to play with this plotting functionality? (like in https://github.com/Project-MONAI/tutorials)

File: .gitignore [hunk 3]
```diff
@@ -157,3 +157,10 @@ cython_debug/
```
Comment (markus-hinsche) [inline .gitignore:157]: The test_data nifty files are 6.2MB each, this is unnecessarily big for the repo. 

I think we should shrink them to smaller sizes, which will:
- reduce repo/git size
- speed up unit tests

File: tests/test_viewer.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+class TestViewer(unittest.TestCase):
+    @parameterized.expand(TEST_CASES)
+    def test_viewers(self, image, label, prediction, description, mode):
+        """Test all viewer in order. Otherwise parallel execution might lead to blocking."""
+        # pyplot windows will open and close during this test in fast succession.
+        # This might lead to the windows blocking each other and the test to fail.
+        # With plt.close("all") and plt.pause(0.001) as well as plt.ion() at the beginning of the file
+        # it seems to work. If tests should fail on another machine, plt.show(block=False) might need to be
+        # added to the viewer classes.
```
[the remark this answers is not in the document]
Comment (kbressem) [reply tests/test_viewer.py:57]: This is the way 😀

I have tried multiple approaches to get the test non blocking. This is why I found after a lot of research. 

Feel free to try out other approaches. But most did not work for me.

File: trainlib/viewer.py [hunk 5]
```diff
@@ -198,23 +214,32 @@ def _create_image_box(self, figsize):
...
 
-    def _generate_views(self):
-        image_box = self._create_image_box(self.figsize)
+    def _ensure_correct_dims(self, array: np.ndarray) -> np.ndarray:
+        array = np.squeeze(array)
+        if array.ndim == 2:
+            array = np.expand_dims(array, 0)  # add channel
+        if array.ndim != 3:
+            raise ValueError(f"Input array needs to be 2d or 3d, but is {array.ndim}d")
+        return array
```
[the remark this answers is not in the document]
Comment (kbressem) [reply trainlib/viewer.py:223]: Everything subclasses or uses BasicViewer. Making it 2D/3D also creates 2D/3D classes for DICOM explorer and for ListViewer and for show_batch. 

I do not think this would be worth it or make the cod more readable.

File: tests/test_viewer.py (hunk 1 above)
Comment (kbressem) [reply tests/test_viewer.py:62]: We need to test the DicomExplorer as well as it can fail differently (and already did). 

I test all possible combinations of 2D/3D with and without labels/masks/predictions and titles. 
Tests for ListViewer could be skipped, as this function only calls BasicViewer.

File: tests/test_trainer.py [hunk 6]
```diff
@@ -64,5 +65,25 @@ def assertPathExists(self, fn: str, is_file: bool = False):  # noqa N802
...
 
+class TestSegmentationTrainer2d(unittest.TestCase):
+    config = deepcopy(TEST_CONFIG_SEGM)
+    config.data.train_csv = config.data.valid_csv.replace(
+        "3d", "2d"
+    )  # train has images with different number of channels
+    config.data.valid_csv = config.data.valid_csv.replace("3d", "2d")
+    config.data.test_csv = config.data.valid_csv.replace("3d", "2d")
+    config.ndim = 2
+    config.data.dataset_type = "iterative"
```
[the remark this answers is not in the document]
Comment (kbressem) [reply tests/test_trainer.py:74]: "iterative" ? This means there is no caching as in PersistentDataset. IterativeDataset is the trainlib (or monai) default dataset.

File: trainlib/utils.py [hunk 7]
```diff
@@ -100,3 +100,9 @@ def import_patched(path: Union[str, Path], name: str) -> Callable:
     function_or_class = getattr(patch, name)
     logger.info(f"importing patch `{name}` from `{path}`.")
     return function_or_class
+
+
+class ShapeMissmatchError(Exception):
```
[the remark this answers is not in the document]
Comment (kbressem) [reply trainlib/utils.py:105]: Errors usually end with `Error`. For example `ValueError`, `TypeError`, `UserError`, therfore I would like to keep the name to make it consistent.

File: tests/test_config_clf.yaml [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+out_dir: output
+overwrite: true
+run_id: runs/name
+seed: 42
+training:
+  early_stopping_patience: 1
+  max_epochs: 1
+transforms:
+  base:
+    LoadImaged:
```
[the remark this answers is not in the document]
Comment (kbressem) [reply tests/test_config_clf.yaml:45]: This is probably correct and should not happen. Thank you for pointing this out.  `get_transform` needs to be adapted for this, which I would like to do in a new PR, where clf-functionality is implemented.

File: tests/test_viewer.py (hunk 2 above)
Comment (kbressem) [reply tests/test_viewer.py:77]: show_batch uses it. Otherwise its just ListViewer([image, image]).show(). What kind of tutorial are you thinking of?

File: setup.cfg [hunk 9]
```diff
@@ -41,5 +41,7 @@ requirements =
 	munch
 	pyyaml
 	bump2version
+	matplotlib
+	ipywidgets
```
[the remark this answers is not in the document]
Comment (kbressem) [reply setup.cfg:45]: I use pretty basic functions from these files and do not expect that the API will be fundametally changed soon. So the biggest risk are probably old versions. 
On the other hand, trainlib will be changed a lot during the next weeks. Wouldn't it be better to pin the libraries to a later timepoint? I'd like to stay close to the newest version.

File: tests/test_viewer.py (hunk 2 above)
Comment (markus-hinsche) [reply tests/test_viewer.py:77]: A super short script that loads the test data and displays it in the different forms. 
This way (reviewer and) potential users can play around with the (interactive) visualizations

Comment (kbressem) [reply tests/test_viewer.py:77]: I've added a notebook, which explains a simple 3d training workflow (as 2d is still buggy). It explains the config and the viewers.

File: .gitignore (hunk 3 above)
Comment (kbressem) [reply .gitignore:157]: Those are already downscaled 3d nifti files.  I'll try adding them to git LFS in the future, but at the moment I am at the end of my quota. 
Given that I/O is done by highly optimized C-classes, unit tests are not much faster with smaller files.

Comment (kbressem) [reply .gitignore:157]: Also, wouldn't the git history stay the same, even if I resize the files now? So the repo will remain the same size (possibly even grow) if I resize the images?

**Comment by kbressem** [comment]
Ok ... this got a bit out of hand yesterday. I decided to test 2d segmentation on another project first and ran into multiple errors which lead to many changes in the library. However, these changes are not necessarily relevant to 2d. As short summary: 

- patching has been changed. The patch file now provides the class not the getter function. This means instead of adding `get_loss` to your `patch_loss.py` you now add, e.g. `My CrossEntropyLoss` to the file, and reference it in the config as you do with every other loss. If find this much easier to use and it gives more flexibility for trying out new losses/models etc. In the future, I might move `get_loss`, `get_model`, `get_optimizer` to utils so custom losses/models/optimizers can be added to trainlib. 
- A new example notebook for 3d segmentation training was added alongside figures. 
- multiple small fixes in trainer/dataloader which previously went unrecognized.

File: examples/example_segmentation_3d.ipynb [hunk 10]
```diff
@@ -0,0 +1,482 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`BasicViewer` can show a single image, with mask overlay or classification label. An optional description can also be added to the plot. \n",
+    "\n",
+    "```python\n",
+    "BasicViewer(image_1, label_1, description=\"A CT showing COVID Pneumonia\").show()\n",
```
Comment (markus-hinsche) [inline examples/example_segmentation_3d.ipynb:252]: When click on `mask`, nothing happens. 

Maybe the mask should be something else than the image (is it the same currently)?

File: examples/example_segmentation_3d.ipynb [hunk 11]
```diff
@@ -0,0 +1,482 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The `DicomExplorer` class provides additional information about the pixel/voxel values in the image array. \n",
+    "\n",
+    "```python\n",
+    "DicomExplorer(image_1, label_1, description=\"A CT showing COVID Pneumonia\").show()\n",
```
Comment (markus-hinsche) [inline examples/example_segmentation_3d.ipynb:264]: I would prefer code in code cells and explanations in markdown cells. 
If I find code in markdown cells, I have to copy-paste (and have local changes) in order to try it

File: examples/example_segmentation_3d.ipynb [hunk 12]
```diff
@@ -0,0 +1,482 @@
+{
```
Comment (markus-hinsche) [inline examples/example_segmentation_3d.ipynb:1]: The dir `examples/` is inconsistent with other repos where we use `notebooks/`.

**Review by markus-hinsche — COMMENTED** [review]
I quickly tried the notebooks and can now see how the 3 different viewers work for 3D. The PR claims to be implementing something for 2d. 
Would it be reasonable to add a notebook or screenshots/docs for 2d too?
